lib/classes/PluginManagement.js

```diff
@@ -0,0 +1,107 @@
+'use strict';
+
+const Utils = require('./Utils');
+const path = require('path');
+const has = require('lodash').has;
+const forEach = require('lodash').forEach;
+
+class PluginManagement {
+  constructor(serverless) {
+    this._serverless = serverless;
+    this._pluginInstances = [];
+    this._commandsList = [];
+    this._commands = {};
+  }
+
+  loadAllPlugins() {
+    this._loadCorePlugins();
+    // this._loadServicePlugins(); // TODO: uncomment once implemented
+  }
+
+  runCommand(command) {
+    const events = this._getEvents(command.split(' '), this._commands);
+    // collect all relevant hooks
+    let hooks = [];
+    events.forEach((event) => {
+      const hooksForEvent = [];
+      this._pluginInstances.forEach((pluginInstance) => {
+        forEach(pluginInstance.hooks, (hook, hookKey) => {
+          if (hookKey === event) {
+            hooksForEvent.push(hook);
+          }
+        });
+      });
+      hooks = hooks.concat(hooksForEvent);
+    });
+
+    // run all relevant hooks one after another
+    // TODO: this code needs to be a bit more complex to support async hooks using Promises
+    hooks.forEach((hook) => {
+      hook();
+    });
+  }
+
+  _addPlugin(Plugin) {
+    this._pluginInstances.push(new Plugin());
+
+    this._loadCommands(Plugin);
+  }
+
+  _loadCorePlugins() {
+    const utils = new Utils();
+    const pluginsDirectoryPath = path.join(__dirname, '../plugins');
+
+    const corePlugins = utils.readFileSync(path.join(pluginsDirectoryPath, 'Plugins.json')).plugins;
+
+    corePlugins.forEach((corePlugin) => {
+      // remove the .js from the plugin name to get the name of the corresponding directory
+      const pluginDirectoryName = corePlugin.slice(0, -3);
+
+      const Plugin = require(path.join(pluginsDirectoryPath, pluginDirectoryName, corePlugin));
+
+      this._addPlugin(Plugin);
+    });
+  }
+
+  _loadServicePlugins() {
+    // TODO implement
+  }
+
+  _loadCommands(Plugin) {
+    this._commandsList.push((new Plugin()).commands);
+
+    // TODO: refactor ASAP as it slows down overall performance
+    // rebuild the commands
+    forEach(this._commandsList, (commands) => {
+      forEach(commands, (commandDetails, command) => {
+        this._commands[command] = commandDetails;
+      });
+    });
+  }
+
+  _getEvents(commandsArray, availableCommands, prefix) {
+    prefix = (typeof prefix !== 'undefined' ? prefix : '');
+    const commandPart = commandsArray[0];
+
+    if (has(availableCommands, commandPart)) {
+      const commandDetails = availableCommands[commandPart];
+      if (commandsArray.length === 1) {
+        const events = [];
+        commandDetails.lifeCycleEvents.forEach((event) => {
+          events.push(`${prefix}${commandPart}:${event}Pre`);
```

Comment:
We should discuss how we can add the `before` here.

The problem is that a prepended before should uppercase the function name but only when a `before` or `after` is prepended:

```
deploy:beforeFunctions
deploy:functions // this is the only correct function name
deploy:afterFunctions
```

I would like to keep the prepending / append in the function name as it reads easier and is easier to implement rather than something like this:

```
before:deploy:function
deploy:function
after:deploy:function
```